tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,206 @@
+# Stateful tests for arbitrary Zarr stores.
+
+import asyncio
+
+import hypothesis.strategies as st
+from hypothesis import note
+from hypothesis.stateful import (
+    RuleBasedStateMachine,
+    invariant,
+    precondition,
+    rule,
+)
+
+from zarr.buffer import Buffer, default_buffer_prototype
+from zarr.store import MemoryStore
+
+# from strategies_store import StoreStatefulStrategies, key_ranges
+from zarr.strategies import key_ranges, paths
+
+# zarr spec: https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html
+
+
+# NOTE: all methods should be called on self.store, assertions should compare self.model and self.store._store_dict
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    def __init__(self):  # look into using run_machine_as_test()
+        super().__init__()
+        self.model = {}
+        self.store = MemoryStore(mode="w")
+
+    # Unfortunately, hypothesis' stateful testing infra does not support asyncio
+    # So we redefine sync versions of the Store API.
+    # https://github.com/HypothesisWorks/hypothesis/issues/3712#issuecomment-1668999041
+    async def store_set(self, key, data_buffer):
+        await self.store.set(key, data_buffer)
```

Comment:
This stuff is independent of the test so lets move it out to a separate class.

```python
class SyncStoreWrapper():
    def __init__(self, store):
        self.store = store
	...
```

Reply:
okay, thanks, and then pass `SyncStoreWrapper()` to `ZarrStoreStateMachine` ? I think that means we'll definitely need to switch to using `run_machine_as_test()` (which we'd need to do anyway to handle different stores )

Reply:
For a start you can just use `SyncStoreWrapper(MemoryStore(mode="w"))` . We can generalize later.